sycl/doc/extensions/proposed/sycl_ext_oneapi_get_kernel_info.asciidoc

```diff
@@ -0,0 +1,228 @@
+----
+
+_Remarks:_ Each information descriptor may specify additional preconditions,
+exceptions that are thrown, etc.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template<typename KernelName, typename Param>
+typename Param::return_type get_kernel_info(const context& ctxt,
+                                            const device& dev);
+
+} // namespace sycl::ext::oneapi
+----
+!====
+
+_Constraints:_ Available only if `Param` is an information descriptor for the
+`kernel` class, which can be used by the `kernel::get_info(const device &)`
+overload.
+
+_Preconditions:_ The `KernelName` must be the type kernel name of a kernel that
+is defined in the application.
+The device `dev` must be one of the devices contained by `ctxt`.
+The kernel `KernelName` must be compatible with the device `dev` as defined by
+`is_compatible`.
+
+_Returns:_ The same value `ret` that would be computed by:
+
+[source,c++]
+----
+auto bundle =
+  sycl::get_kernel_bundle<KernelName, sycl::bundle_state::executable>(ctxt);
+auto ret = bundle.get_kernel<KernelName>().get_info<Param>(dev);
+----
+
+_Remarks:_ Each information descriptor may specify additional preconditions,
+exceptions that are thrown, etc.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+namespace sycl::ext::oneapi {
+
+template<typename KernelName, typename Param>
+typename Param::return_type get_kernel_info(const queue& q);
+
+} // namespace sycl::ext::oneapi
+----
+!====
+
+_Constraints:_ Available only if `Param` is an information descriptor for the
+`kernel` class, which can be used by the `kernel::get_info(const device &)`
+overload.
+
+_Preconditions:_ The `KernelName` must be the type kernel name of a kernel that
+is defined in the application.
+The kernel `KernelName` must be compatible with the device associated with `q`
+as defined by `is_compatible`.
+
+_Returns:_ The same value `ret` that would be computed by:
+
+[source,c++]
+----
+sycl::context ctxt = q.get_context();
+sycl::device dev = q.get_device();
+auto bundle =
+  sycl::get_kernel_bundle<KernelName, sycl::bundle_state::executable>(ctxt);
+auto ret = bundle.get_kernel<KernelName>().get_info<Param>(dev);
+----
+
+_Remarks:_ Each information descriptor may specify additional preconditions,
+exceptions that are thrown, etc.
+
+
+== Issues
+
+* I purposely reduced the exceptions that are required to be thrown for certain
+  error conditions and instead listed these as preconditions.
+  An implementation can still diagnose these error conditions by throwing an
+  exception, but it is not required.
+  Since these APIs are likely on the critical path for launching a kernel, I
+  don't think we want to mandate an error check at runtime.
+  In retrospect, I think this is the right behavior for the core SYCL spec also,
+  and we should consider changing the specified behavior.
+  Thoughts?
```

Comment:
Which exceptions are these, specifically?

Reply:
For example:

> _Preconditions:_ [...] The device `dev` must be one of the devices contained by `ctxt`. The kernel `KernelName` must be compatible with the device `dev` as defined by `is_compatible`.

In the existing SYCL 2020 spec, these error conditions mandate an exception:

> * An `exception` with the `errc::invalid` error code if any of the devices in `devs` is not one of devices contained by the context `ctxt` or is not a descendent device of some device in `ctxt`.
>
> * An `exception` with the `errc::invalid` error code if the kernel is not compatible with device `dev` (as defined by `is_compatible()`)

Reply:
Actually enqueuing the kernel would still throw one of these exceptions though, right?

Reply:
This PR only adds query functions, so nothing in this PR would change the behavior when enqueuing a kernel.

My comment about "in retrospect", though, might change the behavior when enqueuing a kernel.  I do feel that we went too far in the core SYCL spec about mandating exceptions for error conditions.  This requires an implementation to check the function inputs and throw a specific exception.  I'm not sure we want to mandate those error checks, especially for functions that are likely to be on an application's critical path.  If we did not mandate the error checks, an implementation might just assume they are correct, and you get what you get if they are not.  This might be a confusing error from a lower-level software layer, a crash, or even code that runs incorrectly.

Of course, an implementation *could* still perform these error checks.  DPC++ could decide to do this by default if the cost isn't high.  Or, DPC++ could provide a compiler option to enable them if they are expensive.

Reply:
> Of course, an implementation could still perform these error checks. DPC++ could decide to do this by default if the cost isn't high. Or, DPC++ could provide a compiler option to enable them if they are expensive.

Are we allowed to go the other way?

There are already some cases (e.g., math) where a user can opt-in to disabling features required by the specification (e.g., certain NaN handling) in pursuit of performance.  As long as DPC++ provides a mode that is conformant to the required errors, is it legal to provide something like an `-fsycl-no-exceptions` option (and maybe even enable it by default)?

If what I describe above is legal, I think I'd prefer it to adding more UB into the specification.  Developers are already used to implementations providing options that improve performance, and specifying required exceptions guarantees that any conformant implementation will provide some configuration in which it's possible to get the safe behavior.

Reply:
I think it would be possible, but it seems contrary to the way C++ is specified.  Most C++ library functions have a precondition that their arguments are correct.  It's not common for C++ to mandate an error check.